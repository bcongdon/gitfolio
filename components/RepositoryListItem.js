import React from "react";
import { Text, Right, Button, ListItem, Left, Icon } from "native-base";
import moment from "moment";
import nodeEmoji from "node-emoji";
import GitHubColors from "github-colors";

/*
  RepositoriesView is the component that displays a users list of public repositories
 */
export default class RepositoriesListItem extends React.Component {
  componentWillMount() {
    this.setState({ starred: this.props.viewerHasStarred });
  }

  /**
   * Renders an icon tag for a repository
   * @param  {Object} iconOpts Options object to apply to the icon
   * @param  {String} text     Text to display next to the icon
   */
  getRepoIconTag(iconOpts, text) {
    return (
      <Text>
        <Icon {...iconOpts} /> {text}
      </Text>
    );
  }

  /**
   * Renders an icon tag for the provided language
   * @param  {String} language The language
   */
  getLanguageTag(language) {
    const color = GitHubColors.get(language);
    if (!language || !color) {
      return null;
    }

    return this.getRepoIconTag(
      {
        style: { color: color.color, fontSize: 16 },
        type: "MaterialCommunityIcons",
        name: "checkbox-blank-circle"
      },
      language
    );
  }

  /**
   * Renders an icon tag with the provided number of stars
   * @param  {number} stars The number of stars
   */
  getStarsTag(stars) {
    return this.getRepoIconTag(
      {
        style: { fontSize: 16 },
        type: "Octicons",
        name: "star",
        ios: "star"
      },
      stars
    );
  }

  /**
   * Renders an icon tag with the provided number of forks
   * @param  {number} forks The number of forks
   */
  getForksTag(forks) {
    return this.getRepoIconTag(
      {
        style: { fontSize: 16 },
        type: "Octicons",
        name: "repo-forked",
        ios: "repo-forked"
      },
      forks
    );
  }

  /**
   * Renders an icon tag with the the given updated date
   * @param  {Date} pushed_at The "updated_at" date of the repo
   */
  getUpdatedTag(pushed_at) {
    return this.getRepoIconTag(
      {
        style: { fontSize: 16 },
        type: "Octicons",
        name: "calendar",
        ios: "calendar"
      },
      moment(pushed_at).fromNow()
    );
  }

  /**
   * Renders the "star" button for the repo modal
   * @param  {boolean} starred Whether or not the user has starred this repo
   */
  getStarButton(starred, onPress) {
    let buttonType = starred ? "star" : "star-o";
    return (
      <Button transparent dark onPress={onPress}>
        <Icon
          style={{ color: "#fee067" }}
          type="FontAwesome"
          name={buttonType}
          ios={buttonType}
        />
      </Button>
    );
  }

  /**
   * Toggle whether or not the user is starring the current repo
   */
  toggleRepoStar() {
    this.props.onSetStarred(!this.state.starred);
    this.setState({ starred: !this.state.starred });
  }

  /**
   * Renders a repository as a ListItem
   */
  render() {
    // Conditionally render fields as necessary
    let languageTag = this.props.language
      ? this.getLanguageTag(this.props.language)
      : null;
    let starsTag =
      this.props.stars > 0 ? this.getStarsTag(this.props.stars) : null;
    let forksTag =
      this.props.forks > 0 ? this.getForksTag(this.props.forks) : null;

    return (
      <ListItem button onPress={() => this.props.onPress()}>
        <Left>
          <Text>
            {this.props.name}
            {"\n"}

            <Text note>{nodeEmoji.emojify(this.props.description)}</Text>
            {"\n\n"}

            {languageTag}
            {languageTag ? "  " : null}

            {starsTag}
            {starsTag ? "  " : null}

            {forksTag}
            {forksTag ? "  " : null}

            {this.getUpdatedTag(this.props.pushedAt)}
          </Text>
        </Left>
        <Right>
          {this.getStarButton(this.state.starred, () => this.toggleRepoStar())}
        </Right>
      </ListItem>
    );
  }
}
