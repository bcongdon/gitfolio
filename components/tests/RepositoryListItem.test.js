import { ListItem } from "native-base";
import RepositoryListItem from "../RepositoryListItem";
import React from "react";

import renderer from "react-test-renderer";

it("renders without crashing", () => {
  let component = (
    <RepositoryListItem
      name="repo"
      language="Python"
      description="a repo"
      stars={2}
      forks={10}
      viewerHasStarred={false}
    />
  );

  const rendered = renderer.create(component).toJSON();
  expect(rendered).toBeTruthy();
});

it("should snapshot correctly with forks and stars", () => {
  let component = (
    <RepositoryListItem
      name="repo"
      language="Python"
      description="a repo"
      stars={2}
      forks={10}
      viewerHasStarred={false}
    />
  );

  const rendered = renderer.create(component).toJSON();
  expect(rendered).toMatchSnapshot();
});

it("should snapshot correctly without stars", () => {
  let component = (
    <RepositoryListItem
      name="repo"
      language="Python"
      description="a repo"
      stars={0}
      forks={10}
      viewerHasStarred={false}
    />
  );

  const rendered = renderer.create(component).toJSON();
  expect(rendered).toMatchSnapshot();
});

it("should snapshot correctly without forks", () => {
  let component = (
    <RepositoryListItem
      name="repo"
      language="Python"
      description="a repo"
      stars={2}
      forks={10}
      viewerHasStarred={false}
    />
  );

  const rendered = renderer.create(component).toJSON();
  expect(rendered).toMatchSnapshot();
});

it("calls onPress when pressed", () => {
  const onPress = jest.fn();
  let component = (
    <RepositoryListItem
      name="repo"
      language="Python"
      description="a repo"
      stars={2}
      forks={10}
      viewerHasStarred={false}
      url={"repo.example.com"}
      onPress={onPress}
    />
  );

  const rendered = renderer.create(component).root;
  const listItem = rendered.findByType(ListItem);
  listItem.props.onPress();
  expect(onPress).toBeCalled();
});
