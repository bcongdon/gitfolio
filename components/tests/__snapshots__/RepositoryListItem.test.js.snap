// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should snapshot correctly with forks and stars 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={null}
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    button={true}
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderBottomWidth": 0.5,
        "borderColor": "#c9c9c9",
        "flexDirection": "row",
        "marginLeft": 16,
        "paddingBottom": 13,
        "paddingRight": 16,
        "paddingTop": 13,
      }
    }
    virtual={undefined}
  >
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "alignSelf": "center",
          "flex": 1,
          "flexDirection": "row",
        }
      }
      virtual={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        repo
        

        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          note={true}
          style={
            Object {
              "color": "#a7a7a7",
              "fontFamily": "System",
              "fontSize": 14,
            }
          }
          uppercase={false}
          virtual={undefined}
        >
          a repo
        </Text>
        


        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
            }
          }
          uppercase={false}
          virtual={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          />
           
          Python
        </Text>
          
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
            }
          }
          uppercase={false}
          virtual={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          />
           
          2
        </Text>
          
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
            }
          }
          uppercase={false}
          virtual={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          />
           
          10
        </Text>
          
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
            }
          }
          uppercase={false}
          virtual={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          />
           
          a few seconds ago
        </Text>
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "alignSelf": "center",
          "flex": 0.28,
          "paddingBottom": null,
          "paddingLeft": null,
          "paddingRight": null,
          "paddingTop": null,
        }
      }
      virtual={undefined}
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "alignSelf": null,
            "backgroundColor": null,
            "borderBottomWidth": null,
            "borderColor": "#007aff",
            "borderLeftWidth": null,
            "borderRadius": 5,
            "borderRightWidth": null,
            "borderTopWidth": null,
            "elevation": 0,
            "flexDirection": "row",
            "height": 45,
            "justifyContent": "space-between",
            "opacity": 1,
            "paddingBottom": 6,
            "paddingTop": 6,
            "shadowColor": null,
            "shadowOffset": null,
            "shadowOpacity": null,
            "shadowRadius": null,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`should snapshot correctly without forks 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={null}
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    button={true}
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderBottomWidth": 0.5,
        "borderColor": "#c9c9c9",
        "flexDirection": "row",
        "marginLeft": 16,
        "paddingBottom": 13,
        "paddingRight": 16,
        "paddingTop": 13,
      }
    }
    virtual={undefined}
  >
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "alignSelf": "center",
          "flex": 1,
          "flexDirection": "row",
        }
      }
      virtual={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        repo
        

        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          note={true}
          style={
            Object {
              "color": "#a7a7a7",
              "fontFamily": "System",
              "fontSize": 14,
            }
          }
          uppercase={false}
          virtual={undefined}
        >
          a repo
        </Text>
        


        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
            }
          }
          uppercase={false}
          virtual={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          />
           
          Python
        </Text>
          
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
            }
          }
          uppercase={false}
          virtual={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          />
           
          2
        </Text>
          
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
            }
          }
          uppercase={false}
          virtual={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          />
           
          10
        </Text>
          
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
            }
          }
          uppercase={false}
          virtual={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          />
           
          a few seconds ago
        </Text>
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "alignSelf": "center",
          "flex": 0.28,
          "paddingBottom": null,
          "paddingLeft": null,
          "paddingRight": null,
          "paddingTop": null,
        }
      }
      virtual={undefined}
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "alignSelf": null,
            "backgroundColor": null,
            "borderBottomWidth": null,
            "borderColor": "#007aff",
            "borderLeftWidth": null,
            "borderRadius": 5,
            "borderRightWidth": null,
            "borderTopWidth": null,
            "elevation": 0,
            "flexDirection": "row",
            "height": 45,
            "justifyContent": "space-between",
            "opacity": 1,
            "paddingBottom": 6,
            "paddingTop": 6,
            "shadowColor": null,
            "shadowOffset": null,
            "shadowOpacity": null,
            "shadowRadius": null,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`should snapshot correctly without stars 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={null}
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    button={true}
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderBottomWidth": 0.5,
        "borderColor": "#c9c9c9",
        "flexDirection": "row",
        "marginLeft": 16,
        "paddingBottom": 13,
        "paddingRight": 16,
        "paddingTop": 13,
      }
    }
    virtual={undefined}
  >
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "alignSelf": "center",
          "flex": 1,
          "flexDirection": "row",
        }
      }
      virtual={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        repo
        

        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          note={true}
          style={
            Object {
              "color": "#a7a7a7",
              "fontFamily": "System",
              "fontSize": 14,
            }
          }
          uppercase={false}
          virtual={undefined}
        >
          a repo
        </Text>
        


        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
            }
          }
          uppercase={false}
          virtual={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          />
           
          Python
        </Text>
          
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
            }
          }
          uppercase={false}
          virtual={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          />
           
          10
        </Text>
          
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
            }
          }
          uppercase={false}
          virtual={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          />
           
          a few seconds ago
        </Text>
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "alignSelf": "center",
          "flex": 0.28,
          "paddingBottom": null,
          "paddingLeft": null,
          "paddingRight": null,
          "paddingTop": null,
        }
      }
      virtual={undefined}
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "alignSelf": null,
            "backgroundColor": null,
            "borderBottomWidth": null,
            "borderColor": "#007aff",
            "borderLeftWidth": null,
            "borderRadius": 5,
            "borderRightWidth": null,
            "borderTopWidth": null,
            "elevation": 0,
            "flexDirection": "row",
            "height": 45,
            "justifyContent": "space-between",
            "opacity": 1,
            "paddingBottom": 6,
            "paddingTop": 6,
            "shadowColor": null,
            "shadowOffset": null,
            "shadowOpacity": null,
            "shadowRadius": null,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
      </View>
    </View>
  </View>
</View>
`;
